common/src/main/java/org/apache/iceberg/common/testutils/PerTestCallbackWrapper.java

```diff
@@ -0,0 +1,51 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.common.testutils;
+
+import org.junit.jupiter.api.extension.AfterEachCallback;
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * This class is used to wrap a specific {@link CustomExtension} to junit5 extension with test-level
+ * callbacks {@link BeforeEachCallback} and {@link AfterEachCallback}. It can be used as a
+ * substitute for @Rule in junit4.
+ */
+public class PerTestCallbackWrapper<C extends CustomExtension>
```

Comment:
why do we need `PerTestCallbackWrapper` and `PerClassCallbackWrapper`? I think all of the logic should just go into `EcsS3MockExtension` and only implement `BeforeAllCallback, AfterAllCallback, BeforeEachCallback, AfterEachCallback`. Also I don't see a reason for having `CustomExtension`

Reply:
If the `EcsS3MockExtension` implements both `BeforeEachCallback ` and `BeforeAllCallback `, then the lifecycle method like `initialize` of `EcsS3MockRule ` will be invoked double times.  Junit4 `TestRule` can take care of this by annotation with `@Rule` or `@ClassRule` respectively. As for Junit5, we can only register extension with single annotation `@RegisterExtension`. `PerTestCallbackWrapper ` and `PerClassCallbackWrapper ` is used to control the lifecycle of `initialize`(`before all` or `before each`).

The core is to implement an `CustomExtension` similar to `TestRule`, and then select the appropriate `PerTest/PerClassCallBackWrapper` to wrap it based on the actual life cycle required for testing.

The reason for introducing `CustomExtension` is for future extends. Of course, we may only need to migrate this `EcsS3TestRule` to Junit5 temporarily, but I think it is worth it. 🤔

Reply:
Agree with @nastra here. `EcsS3MockExtension` should be the only JUnit extension. JUnit allows injection - for static fields, instance fields and test parameters. Also, `ExtensionContext` in JUnit is hierarchical (root, ..., test class, test instance, test method, etc).

BTW test code must not end in production ("main") code.

Reply:
Thank you for your reply, but I still have some doubts:

If we implement both `beforeAllCallback` and `beforeEachCallback` for `EcsS3MockExtension` and register it using `static` field, then both `beforeEach` and `beforeAll` will be invoked(`class level` and `method level` respectively). If it is registered using a `non-static field`, only `beforeEach` will be called at the `method level`. 

Is my usage incorrect? Is there any way we can avoid calling `method level` methods during static registration (which corresponds to `@ClassRule` in junit4).

Reply:
I think, your understanding is not correct.
The JUnit docs have good information about how that works: https://junit.org/junit5/docs/current/user-guide/#extensions-registration

Reply:
I just writing an example to verify this as following: 

Define the extension:
```
public class TestExtension
    implements BeforeAllCallback, AfterAllCallback, BeforeEachCallback, AfterEachCallback {

  public void afterAll(ExtensionContext context) throws Exception {
    System.out.println("after all");
  }

  public void afterEach(ExtensionContext context) throws Exception {
    System.out.println("after each");
  }

  public void beforeAll(ExtensionContext context) throws Exception {
    System.out.println("before all");
  }

  @Override
  public void beforeEach(ExtensionContext context) throws Exception {
    System.out.println("before each");
  }
}
```

Test Class:
```
class ExtensionTest {
  @RegisterExtension static TestExtension testExtension = new TestExtension();

  @Test
  void test1() {
    System.out.println("test1");
  }

  @Test
  void test2() {
    System.out.println("test2");
  }
}
```

The output is:
```
before all
before each
test1
after each
before each
test2
after each
after all
```

And If I directly use the `ExtendWith` annotation instead of static `RegisterExtension`, the output is also the same.
```
@ExtendWith(TestExtension.class)
class ExtensionTest {}
```